### PR TITLE
Reconfigure CI pipelines for PR and merge queue

### PR DESCRIPTION
**Pull request by jakelishman**
### Summary

This commit refactors the pipeline organisation of our CI to reduce the total number of jobs run on any given PR merge, and to reduce the critical path length of merging.  In particular:

- the lint, docs and QPY tests are combined into one job for a single CI worker.  These three jobs individually add up to less than a test run, and moving QPY from the first-stage PR test run to this rebalances the (now) two jobs in the stage to be more equal in runtime.

- a new pipeline is added specifically for the merge queue.  Previously this reused the same two-stage PR pipeline.  The two-stage system unnecessarily lengthened the critical path, as when a PR enters the merge queue, it as already passed PR CI, and so is highly likely to pass all jobs.  In addition to flattening to a single stage, one each of the macOS and Windows jobs are removed to lower the amount of VMs needed and reduce the chances of a timeout (these OSes are more likely to get a dodgy VM and time out than Linux).  The only way a new test failure should appear (other than a flaky test) is by a logical merge conflict, which would be quite unlikely to only affect a particular Python version.

To make it easier to run the lint and docs jobs together, and ensure that both run even if there's a failure in one, the full lint configuration is merged back to `tox.ini`, and that's used to do the linting.  This makes it more consistent for developers, as well.

### Details and comments

This should automatically already work with the branch-protection rules, but there's like a 99% chance I've messed something up in the Azure configuration on the first attempt, so let's see if CI passes.

### CODE REVIEW

File: .azure/lint_docs_qpy-linux.yml [hunk 1]
```diff
@@ -20,14 +21,13 @@ jobs:
...
-        displayName: 'Install dependencies'
+        displayName: 'Install docs dependencies'
 
-      - bash: |
-          set -e
-          tox -e docs
-          # Clean up Sphinx detritus.
-          rm -rf docs/_build/html/{.doctrees,.buildinfo}
-        displayName: 'Run Docs build'
+      - bash: tox run -e docs,lint
```
Comment (mtreinish) [inline .azure/lint_docs_qpy-linux.yml:26]: Oh good idea to leverage tox for lint too. Save a lot of config logic here.

File: tox.ini [hunk 2]
```diff
@@ -23,9 +23,16 @@ commands =
 
 [testenv:lint]
 basepython = python3
+# `pylint` will examine the source code, not the version that would otherwise be
+# installed in `site-packages`, so we use an editable install to make sure the
+# compiled modules are built into a valid place for it to find them.
+package = editable
```
Comment (mtreinish) [inline tox.ini:29]: Maybe we should do this globally in the tox config. I'm constantly hitting this locally when using tox which is why I pushed https://github.com/Qiskit/qiskit/pull/11380

**Review by mtreinish — COMMENTED** [review]
Overall this LGTM, just one question inline

File: tox.ini (hunk 2 above)
Comment (jakelishman) [reply tox.ini:29]: Let's merge #11380 in that separate PR after this, maybe? To me, this change is required to fix the lint job (at the moment, `tox -e lint` only works if you happen to already have built the rust extensions in place yourself, hence the CI failure earlier), but #11380 isn't strictly _needed_, it's a change to the config to deliberately override some of tox's isolation as a trade-off to get better re-test performance.

File: .azure/lint_docs_qpy-linux.yml (hunk 1 above)
Comment (jakelishman) [reply .azure/lint_docs_qpy-linux.yml:26]: laziness is often a good motivator for single-source-of-truth configuration lol

**Review by mtreinish — DISMISSED** [review]

**Comment by jakelishman** [comment]
Ha, the templating of the merge queue CI is bust - it crashed when added to the queue.

File: azure-pipelines.yml [hunk 3]
```diff
@@ -193,6 +196,46 @@ stages:
...
 
+  # Merge queue.  A PR that reaches here has already passed the more rigorous PR
+  # suite, so is very likely to pass.  The main reasons for failures here are
+  # flaky VMs timing out (which we can't do much about), or a merge conflict
+  # with another PR that is also in the merge queue.
+  #
+  # There's no reason to have multiple stages in this case, because we're
+  # expecting it to pass.  Having more than one stage frustrates parallel
+  # throughput in low-contention cases, and guarantees a longer critical path.
+  - ${{ if and(eq(variables['Build.Reason'], 'IndividualCI'), startsWith(variables['Build.SourceBranch'], 'refs/heads/gh-merge-queue')) }}:
```
Comment (jakelishman) [inline azure-pipelines.yml]:
```suggestion
  - ${{ if and(eq(variables['Build.Reason'], 'IndividualCI'), contains(variables['Build.SourceBranch'], 'gh-readonly-queue')) }}:
```
duh

**Comment by jakelishman** [comment]
Dunno how I borked that - I'd assumed I'd copy-pasted the condition, but apparently not.  Hopefully better now.

**Review by mtreinish — APPROVED** [review]

**Comment by mtreinish** [comment]
@Mergifyio backport stable/0.46

**Pull request merged**